src/main/java/com/shapesecurity/salvation/directiveValues/HostSource.java

```diff
@@ -128,4 +129,58 @@ public boolean matchesOnlyOrigin(@Nonnull SchemeHostPortTriple origin) {
             (isDefaultPort ? "" : ":" + (this.port == Constants.WILDCARD_PORT ? "*" : this.port)) +
             (this.path == null ? "" : this.path);
     }
+
+    public static boolean pathMatches(@Nullable String pathA, @Nullable String pathB) {
+
+        if (pathA == null || pathA.isEmpty()) {
+            return true;
+        }
+
+        if (pathB == null || pathB.isEmpty()) {
+            return false;
+        }
+
+        if (pathA.matches("/") && pathB.isEmpty()) {
+            return true;
+        }
+
+        boolean exactMatch = pathA.endsWith("/") ? false : true;
+
+        List<String> pathListA = Arrays.asList(pathA.split("/"));
+        List<String> pathListB = Arrays.asList(pathB.split("/"));
+
+        long countA = pathA.chars().filter(ch -> ch =='/').count();
+        long countB = pathB.chars().filter(ch -> ch =='/').count();
+
+        if (pathListA.size() > pathListB.size() || countA > countB)
+            return false;
+
+        if (exactMatch && (pathListA.size() != pathListB.size() || countA != countB)) {
+            return false;
+        }
+
+        StringTokenizer st = new StringTokenizer(pathA, "/");
+        int z = st.countTokens();
+
+        Iterator it1 = pathListA.iterator();
+        Iterator it2 = pathListB.iterator();
+
+        while (it1.hasNext() && it2.hasNext()) {
+            String a = decodeString((String) it1.next());
+            String b = decodeString((String) it2.next());
+            if (!a.equals(b)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    public static String decodeString(@Nonnull String s) {
+        String out = "";
```

Comment:
I might implement this function like this:

```java
public static String decodeString(@Nonnull String s) {
    try {
        return URLDecoder.decode(s, "UTF-8");
    } catch (Exception e) {
        return s;
    }
}
```

The reason is that illegal sequences such as `%%` or `%0G` would probably cause this to throw, but should be preserved as-is instead. We should have tests. My concern with the implementation I've suggested is that a path part containing both illegal sequences and legal sequences should still have the replacement done on the legal sequences. We may have to come up with a custom solution.